src/Nethermind/Nethermind.Xdc/EpochSwitchManager.cs

```diff
@@ -0,0 +1,454 @@
+            {
+                return null;
+            }
+        }
+
+        return epochSwitchInfo;
+    }
+
+    internal bool TryGetBlockByEpochNumber(ulong tempTCEpoch, out Types.BlockRoundInfo epochBlockInfo)
+    {
+        var headHeader = _tree.Head.Header as XdcBlockHeader;
+
+        var xdcSpec = _xdcSpecProvider.GetXdcSpec(headHeader);
+
+        EpochSwitchInfo epochSwitchInfo = GetEpochSwitchInfo(headHeader);
+        if (epochSwitchInfo is null)
+        {
+            epochBlockInfo = null;
+            return false;
+        }
+
+        ulong epochRound = (ulong)xdcSpec.SwitchBlock + epochSwitchInfo.EpochSwitchBlockInfo.Round / (ulong)xdcSpec.EpochLength;
+
+        if (tempTCEpoch == epochRound)
+        {
+            epochBlockInfo = epochSwitchInfo.EpochSwitchBlockInfo;
+            return true;
+        }
+
+        if (tempTCEpoch > epochRound)
+        {
+            epochBlockInfo = null;
+            return false;
+        }
+
+        if (tempTCEpoch < (ulong)xdcSpec.SwitchEpoch)
+        {
+            epochBlockInfo = null;
+            return false;
+        }
+
+        ulong estRound = (tempTCEpoch - (ulong)xdcSpec.SwitchEpoch) * (ulong)xdcSpec.EpochLength;
+
+        if (TryGetBlockInfoInCache(estRound, (ulong)xdcSpec.EpochLength, out epochBlockInfo))
+        {
+            return true;
+        }
+
+        var epoch = (ulong)xdcSpec.EpochLength;
+        ulong estBlockNumDiff = epoch * (epochRound - tempTCEpoch);
+        long estBlockNum = epochSwitchInfo.EpochSwitchBlockInfo.BlockNumber - (long)estBlockNumDiff;
+
+        if (estBlockNum < xdcSpec.SwitchBlock)
+        {
+            estBlockNum = (long)xdcSpec.SwitchBlock;
+        }
+
+        ulong closeEpochNum = 2ul;
+
+        if (closeEpochNum >= epochRound - tempTCEpoch)
+        {
+            var estBlockHeader = (XdcBlockHeader)_tree.FindHeader(estBlockNum);
+            if (estBlockHeader is null)
+            {
+                epochBlockInfo = null;
+                return false;
+            }
+            var epochSwitchInfos = GetEpochSwitchBetween(estBlockHeader, headHeader);
+            if (epochSwitchInfos is null)
+            {
+                return false;
+            }
+            foreach (var info in epochSwitchInfos)
+            {
+                ulong epochNum = (ulong)xdcSpec.SwitchEpoch + info.EpochSwitchBlockInfo.Round / (ulong)xdcSpec.EpochLength;
+                if (epochNum == tempTCEpoch)
+                {
+                    epochBlockInfo = info.EpochSwitchBlockInfo;
+                    return true;
+                }
+            }
+        }
+
+        return TryBinarySearchBlockByEpochNumber(tempTCEpoch, estBlockNum, epochSwitchInfo.EpochSwitchBlockInfo.BlockNumber, (ulong)xdcSpec.SwitchBlock, (ulong)xdcSpec.EpochLength, out epochBlockInfo);
+    }
+
+    private bool TryGetBlockInfoInCache(ulong estRound, ulong epoch, out BlockInfo epochBlockInfo)
```

Comment:
Same reasoning with this one, I would just leave it as `private BlockInfo? GetBlockInfoInCache(ulong estRound, ulong epoch)`